### PR TITLE
grpc instrumentation close spans in halfClose #862

### LINKED ISSUE #862: grpc instrumentation error
When an error occurs in a grpc service, the span is not finished correctly.

It's related with this issue: https://github.com/grpc/grpc-java/issues/1552

The solution could be something like this, in class `brave.grpc.TracingServerInterceptor#ScopingServerCallListener`:

```
@Override public void onHalfClose() {
      SpanInScope scope = tracer.withSpanInScope(span);
      try { // retrolambda can't resolve this try/finally
          delegate().onHalfClose();
      } catch (RuntimeException | Error e) {
          span.error(e);
          span.finish();
          throw e;
      }  finally {
        scope.close();
      }
}
```

### PR DESCRIPTION
**Pull request by oburgosm**
Fixes #862

### CODE REVIEW

File: instrumentation/grpc/src/main/java/brave/grpc/TracingServerInterceptor.java [hunk 1]
```diff
@@ -136,6 +136,10 @@ public void sendMessage(RespT message) {
       SpanInScope scope = tracer.withSpanInScope(span);
       try { // retrolambda can't resolve this try/finally
         delegate().onHalfClose();
+      } catch (RuntimeException | Error e) {
```
Comment (adriancole) [inline instrumentation/grpc/src/main/java/brave/grpc/TracingServerInterceptor.java]: is the intent to close the span on half-close only on error? maybe add a comment why

Comment (oburgosm) [reply instrumentation/grpc/src/main/java/brave/grpc/TracingServerInterceptor.java]: I have replicated the behavior of the 'close' method, but I think there would be no problem in capturing Throwable instead.

Comment (adriancole) [reply instrumentation/grpc/src/main/java/brave/grpc/TracingServerInterceptor.java]: I guess then if half-close occurs then close wouldn't? sorry I'm not aware of the state diagram.

Comment (adriancole) [reply instrumentation/grpc/src/main/java/brave/grpc/TracingServerInterceptor.java]: So, I suppose a comment would be something like..

```
// if there was an exception executing onHalfClose, we don't expect other lifecycle commands
// Accordingly, we close the span
```

Comment (oburgosm) [reply instrumentation/grpc/src/main/java/brave/grpc/TracingServerInterceptor.java]:
```suggestion
      } catch (RuntimeException | Error e) {
      // if there was an exception executing onHalfClose, we don't expect other lifecycle commands
      // Accordingly, we close the span
```

Comment (adriancole) [reply instrumentation/grpc/src/main/java/brave/grpc/TracingServerInterceptor.java]: I will edit the file directly

**Pull request merged**